crates/render/src/lib.rs

```diff
@@ -0,0 +1,48 @@
+//! Crate that provides wrapper for
+//! [requestAnimationFrame](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestAnimationFrame)
+
+#![deny(missing_docs, missing_debug_implementations)]
+
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::JsCast;
+
+/// Handle for [`request_animation_frame`].
+#[derive(Debug)]
+pub struct AnimationFrame {
+    render_id: i32,
+    closure: Closure<dyn Fn(JsValue)>,
+}
+
+impl Drop for AnimationFrame {
+    fn drop(&mut self) {
+        web_sys::window()
+            .unwrap_throw()
+            .cancel_animation_frame(self.render_id)
+            .unwrap_throw()
+    }
+}
+
+/// Calls browser's `requestAnimationFrame`. It is cancelled when the handler is dropped.
+///
+/// [MDN Documentation](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestAnimationFrame)
+pub fn request_animation_frame<F>(callback: F) -> AnimationFrame
+where
+    F: Fn(f64) + 'static,
+{
+    let callback = move |v: JsValue| {
+        let time: f64 = v.as_f64().unwrap_or(0.0);
```

Comment:
The animation frame callback is only called once right? How about when it's called, we drop the closure and clear the render_id so that the `AnimationFrame` drop implementation becomes a no-op?

Reply:
> Also, trying to drop the callback and render_id early would just add more complexity. I don't understand how we could do that.

Typically it's done using something like `Rc<RwLock<Option<...>>>`. You could then clone the `Rc` into the closure to get a mutable reference to the inner option

Reply:
Right, but I'm not sure if it's worth the overhead considering there has to be code in `Drop` impl. I mentioned this above:

> We would still need to keep the code in drop because what if `AnimationFrame` is dropped before the callback is called.

Reply:
Most times the callback will not be dropped before being called, by skipping the `cancel_animation_frame` we save some relatively slow browser api calls and free up any memory that is captured in the user's callback

Reply:
Done that.